File: src/policy/schema.ts
import { z } from 'zod';

export const GenParams = z.object({
  max_tokens: z.number().optional(),
  ctx: z.number().optional(),
  temperature: z.number().optional(),
  timeout_ms: z.number().optional()
});

export const ModelDef = z.object({
  candidates: z.array(z.string()).optional(),
  candidates_cpu: z.array(z.string()).optional()
}).merge(GenParams);

export const FacetDef = z.object({
  model: z.string(),
  engine: z.enum(['ollama', 'llama_cpp', 'vllm']),
  always_loaded: z.boolean().default(false),
  max_tokens: z.number(),
  ctx: z.number(),
  temperature: z.number()
});

export const Policy = z.object({
  routing: z.object({
    escalate_tags: z.array(z.string()),
    hard_keywords: z.array(z.string()),
    default_class: z.enum(['trivial', 'normal', 'hard'])
  }),
  models: z.object({
    router: ModelDef,
    primary: ModelDef,
    heavy: ModelDef
  }),
  // Task-routed mesh: which specialist model handles which kind of work,
  // independent of the trivial/normal/hard difficulty axis above.
  facets: z.object({
    planner: FacetDef,
    coder: FacetDef,
    fast: FacetDef
  }).optional(),
  autotune: z.object({
    first_token_threshold_ms: z.number(),
    total_threshold_ms: z.number(),
    success_rate_threshold: z.number()
  }),
  policy: z.object({
    mode: z.enum(['dev', 'prod']),
    fallback_enabled: z.boolean(),
    log_routing_decisions: z.boolean()
  }),
  endpoints: z.object({
    vllm_base_url: z.string()
  })
});

export type PolicyConfig = z.infer<typeof Policy>;

File: jarvis/policy/schema.py
"""Policy schema, ported from src/policy/schema.ts (Zod -> Pydantic)."""
from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel


class ModelDef(BaseModel):
    candidates: Optional[list[str]] = None
    candidates_cpu: Optional[list[str]] = None
    max_tokens: Optional[int] = None
    ctx: Optional[int] = None
    temperature: Optional[float] = None
    timeout_ms: Optional[int] = None


class FacetDef(BaseModel):
    """One specialist model in the task-routed local mesh."""

    model: str
    engine: Literal["ollama", "llama_cpp", "vllm"]
    always_loaded: bool = False
    max_tokens: int
    ctx: int
    temperature: float


class RoutingConfig(BaseModel):
    escalate_tags: list[str]
    hard_keywords: list[str]
    default_class: Literal["trivial", "normal", "hard"]


class ModelsConfig(BaseModel):
    router: ModelDef
    primary: ModelDef
    heavy: ModelDef


class FacetsConfig(BaseModel):
    planner: FacetDef
    coder: FacetDef
    fast: FacetDef


class AutotuneConfig(BaseModel):
    first_token_threshold_ms: int
    total_threshold_ms: int
    success_rate_threshold: float


class PolicyMode(BaseModel):
    mode: Literal["dev", "prod"]
    fallback_enabled: bool
    log_routing_decisions: bool


class EndpointsConfig(BaseModel):
    vllm_base_url: str


class PolicyConfig(BaseModel):
    routing: RoutingConfig
    models: ModelsConfig
    # Task-routed mesh: which specialist model handles which kind of work,
    # independent of the trivial/normal/hard difficulty axis above.
    facets: Optional[FacetsConfig] = None
    autotune: AutotuneConfig
    policy: PolicyMode
    endpoints: EndpointsConfig
